### PR TITLE
Switch the REST v2 api endpoints

### PR DESCRIPTION
**Pull request by goldbattle**
Pull request checklist:

- [ ] `CHANGELOG.md` was updated, if applicable

# Description

This switches to using the REST v2 api endpoint instead of the GQL one. I also changed the emote URL logic to use the ones returned from the API to make them "data driven". This was requested as per https://github.com/Chatterino/chatterino2/pull/3071#issuecomment-1008454760 and is part of my effort to upstream 7tv support (see https://github.com/Chatterino/chatterino2/pull/3071#issuecomment-1008367948).

### CODE REVIEW

File: src/providers/seventv/SeventvEmotes.cpp [hunk 1]
```diff
@@ -131,114 +139,44 @@ boost::optional<EmotePtr> SeventvEmotes::emote(const EmoteName &name) const
...
-    payload.insert("query", query.replace(whitespaceRegex, " "));
-    payload.insert("variables", variables);
-
-    qDebug() << QJsonDocument(payload).toJson(QJsonDocument::Compact);
-
-    NetworkRequest(apiUrlGQL, NetworkRequestType::Post)
-        .timeout(20000)
-        .header("Content-Type", "application/json")
-        .payload(QJsonDocument(payload).toJson(QJsonDocument::Compact))
+    NetworkRequest("https://api.7tv.app/v2/users/" + channelDisplayName +
```
Comment (jammehcow) [inline src/providers/seventv/SeventvEmotes.cpp:169]: Why did you choose to use the channel's display name here? Twitch IDs are supported when querying a channel's emotes (e.g. `https://api.7tv.app/v2/users/82674227/emotes` would resolve mine)

File: src/providers/seventv/SeventvEmotes.cpp [hunk 2]
```diff
@@ -131,114 +139,44 @@ boost::optional<EmotePtr> SeventvEmotes::emote(const EmoteName &name) const
...
-
-            auto pair = parseGlobalEmotes(parsedEmotes, *this->global_.get());
+        .onSuccess([this](auto result) -> Outcome {
+            auto emotes = this->emotes();
+            auto pair = parseGlobalEmotes(result.parseJsonArray(), *emotes);
             if (pair.first)
                 this->global_.set(
                     std::make_shared<EmoteMap>(std::move(pair.second)));
             return pair.first;
+            return Failure;
```
Comment (jammehcow) [inline src/providers/seventv/SeventvEmotes.cpp:155]: Unreachable return?

File: src/providers/seventv/SeventvEmotes.cpp [hunk 3]
```diff
@@ -131,114 +139,44 @@ boost::optional<EmotePtr> SeventvEmotes::emote(const EmoteName &name) const
...
-    variables.insert("query", QString());
-    variables.insert("globalState", "only");
-    variables.insert("page", 1);  // TODO(zneix): Add support for pagination
-    variables.insert("limit", maxPageSize);
-    variables.insert("pageSize", maxPageSize);
+    qCDebug(chatterinoSeventv) << "[7TVEmotes] Loading 7TV Global Emotes";
 
-    payload.insert("query", query.replace(whitespaceRegex, " "));
-    payload.insert("variables", variables);
+    QString url("https://api.7tv.app/v2/emotes/global");
```
Comment (jammehcow) [inline src/providers/seventv/SeventvEmotes.cpp:144]: Would suggest moving this URL directly into the `NetworkRequest(...)` call instead of creating a variable as it's never used later on. (GH doesn't allow suggestions across deleted lines, sorry)

**Review by jammehcow — CHANGES_REQUESTED** [review]
Looks good, other than the usage of a channel's display name when calling the 7TV API. Can just use the ID.

File: src/providers/seventv/SeventvEmotes.cpp (hunk 2 above)
Comment (goldbattle) [reply src/providers/seventv/SeventvEmotes.cpp:155]: Fix this, but seems I didn't push before I logged off for the night. This should be removed.

File: src/providers/seventv/SeventvEmotes.cpp (hunk 3 above)
Comment (goldbattle) [reply src/providers/seventv/SeventvEmotes.cpp:144]: Sure, I based this off the BTTV module. I will change there too.

File: src/providers/seventv/SeventvEmotes.cpp (hunk 1 above)
Comment (goldbattle) [reply src/providers/seventv/SeventvEmotes.cpp:169]: Didn't realize there was ID support (only looked at the docs). I will switch it to id so things are cleaner.

File: src/providers/seventv/SeventvEmotes.cpp [hunk 4]
```diff
@@ -131,114 +139,44 @@ boost::optional<EmotePtr> SeventvEmotes::emote(const EmoteName &name) const
...
-            }
-        }
-    })";
-
-    variables.insert("query", QString());
-    variables.insert("globalState", "only");
-    variables.insert("page", 1);  // TODO(zneix): Add support for pagination
-    variables.insert("limit", maxPageSize);
-    variables.insert("pageSize", maxPageSize);
+    qCDebug(chatterinoSeventv) << "[7TVEmotes] Loading 7TV Global Emotes";
```
Comment (zneix) [inline src/providers/seventv/SeventvEmotes.cpp:142]: This is an unnecessary change - logging categories exist for a reason and are there to not hardcode categories like this.

File: src/providers/seventv/SeventvEmotes.cpp [hunk 5]
```diff
@@ -131,114 +139,44 @@ boost::optional<EmotePtr> SeventvEmotes::emote(const EmoteName &name) const
...
-                        id
-                        display_name
-                        login
-                        twitch_id
-                    }
-                }
-            }
-        })";
+    qCDebug(chatterinoSeventv) << "[7TVEmotes] Reloading 7TV Channel Emotes"
+                               << channelId << manualRefresh;
```
Comment (zneix) [inline src/providers/seventv/SeventvEmotes.cpp:167]: This is an unnecessary change - logging categories exist for a reason and are there to not hardcode categories like this.

File: src/providers/seventv/SeventvEmotes.cpp [hunk 6]
```diff
@@ -131,114 +139,44 @@ boost::optional<EmotePtr> SeventvEmotes::emote(const EmoteName &name) const
...
-            auto emoteMap = parseChannelEmotes(parsedEmotes, channelId);
+            auto emoteMap =
+                parseChannelEmotes(result.parseJsonArray(), channelId);
             bool hasEmotes = !emoteMap.empty();
 
             qCDebug(chatterinoSeventv)
-                << "Loaded 7TV Channel Emotes" << channelId << emoteMap.size()
-                << manualRefresh;
+                << "[7TVEmotes] Loaded 7TV Channel Emotes" << channelId
+                << emoteMap.size() << manualRefresh;
```
Comment (zneix) [inline src/providers/seventv/SeventvEmotes.cpp:179]: This is an unnecessary change - logging categories exist for a reason and are there to not hardcode categories like this.

File: src/providers/seventv/SeventvEmotes.cpp [hunk 7]
```diff
@@ -24,12 +24,20 @@ namespace {
...
+        {
+            // 7tv has its emotes in an array for some reason
+            // The first entry in the array is the scale, second is the url
+            auto urlArray = urlPair.toArray();
+            if (urlArray.size() == 2 && urlArray.at(0).toString() == emoteScale)
+            {
+                return {urlArray.at(1).toString()};
+            }
+        }
+        return {""};
```
Comment (zneix) [inline src/providers/seventv/SeventvEmotes.cpp:40]: Scheme for emote URLs hasn't changed, why do we change this function to a very confusing syntax instead of just using emote IDs? Just assuming array entries based on their indexes and then comparing with hardcoded value passed through `emoteScale` seems like a clumsy solution. I'd honestly much rather prefer old solution or alternatively, get rid of this function and make an `ImageSet` consisting of `Image`s from a for loop or something. Maybe I'm too picky, no idea.

File: src/providers/seventv/SeventvEmotes.cpp (hunk 1 above)
Comment (zneix) [reply src/providers/seventv/SeventvEmotes.cpp:169]: Also keep in mind that [`QString::arg()`](https://doc.qt.io/qt-5/qstring.html#arg) should rather be used instead of concatenating strings together like this.
Just noticed that you've removed `apiUrlGQL` from the header file, perhaps a format string should be rather declared and used here with an arg method instead of concatenating paths together. Look at my example here: https://github.com/SevenTV/chatterino7/pull/68/files#diff-ca99a25309ab269b0aa09e72c2f777538409d73926e95d33155c20b730e6bce0L30-L32

File: src/providers/ffz/FfzEmotes.cpp [hunk 8]
```diff
@@ -184,7 +184,6 @@ void FfzEmotes::loadEmotes()
     QString url("https://api.frankerfacez.com/v1/set/global");
 
     NetworkRequest(url)
-
```
Comment (zneix) [inline src/providers/ffz/FfzEmotes.cpp:187]: Unrelated change(?)

File: src/providers/ffz/FfzEmotes.hpp [hunk 9]
```diff
@@ -14,9 +14,6 @@ class EmoteMap;
 
 class FfzEmotes final
 {
-    static constexpr const char *globalEmoteApiUrl =
-        "https://api.frankerfacez.com/v1/set/global";
-
```
Comment (zneix) [inline src/providers/ffz/FfzEmotes.hpp:19]: Unrelated change(?)

File: src/providers/seventv/SeventvEmotes.hpp [hunk 10]
```diff
@@ -43,6 +41,7 @@ class SeventvEmotes final
     void loadEmotes();
     static void loadChannel(std::weak_ptr<Channel> channel,
                             const QString &channelId,
+                            const QString &channelDisplayName,
```
Comment (zneix) [inline src/providers/seventv/SeventvEmotes.hpp:44]: Why do we need display name now? Keep in mind that this can be non-ascii characters in some cases, and if anything - login name should be used instead of display name. Again, I think this is an unnecessary argument.

**Review by zneix — CHANGES_REQUESTED** [review]

File: src/providers/seventv/SeventvEmotes.cpp (hunk 4 above)
Comment (goldbattle) [reply src/providers/seventv/SeventvEmotes.cpp:142]: This was based off the FFZ class, I will change there too.

File: src/providers/seventv/SeventvEmotes.cpp (hunk 7 above)
Comment (goldbattle) [reply src/providers/seventv/SeventvEmotes.cpp:40]: This was requested here https://github.com/Chatterino/chatterino2/pull/3071#issuecomment-1008346507 by @AnatoleAM to future proof it to different url structures. This loop is required since the actual API object isn't a map (it really should be).

File: src/providers/ffz/FfzEmotes.hpp (hunk 9 above)
Comment (goldbattle) [reply src/providers/ffz/FfzEmotes.hpp:19]: This variable is unused.

File: src/providers/seventv/SeventvEmotes.hpp (hunk 10 above)
Comment (goldbattle) [reply src/providers/seventv/SeventvEmotes.hpp:44]: As mentioned above:
> Didn't realize there was ID support (only looked at the docs). I will switch it to id so things are cleaner.

File: src/providers/seventv/SeventvEmotes.cpp (hunk 1 above)
Comment (goldbattle) [reply src/providers/seventv/SeventvEmotes.cpp:169]: Sure, the arg method seems a bit more elegant. I will change to that.

File: src/providers/seventv/SeventvEmotes.cpp (hunk 7 above)
Comment (AnatoleAM) [reply src/providers/seventv/SeventvEmotes.cpp:40]: It has changed to a different structure in API V3 for the record.

File: src/providers/ffz/FfzEmotes.cpp (hunk 8 above)
Comment (jammehcow) [reply src/providers/ffz/FfzEmotes.cpp:187]: Just a whitespace change, fine by me.

File: src/providers/seventv/SeventvEmotes.hpp (hunk 10 above)
Comment (jammehcow) [reply src/providers/seventv/SeventvEmotes.hpp:44]: As long as it's removed with the change back to IDs I'm happy.

File: src/providers/ffz/FfzEmotes.hpp (hunk 9 above)
Comment (jammehcow) [reply src/providers/ffz/FfzEmotes.hpp:19]: Doesn't impact functionality, fine by me.

File: src/providers/seventv/SeventvEmotes.cpp (hunk 4 above)
Comment (zneix) [reply src/providers/seventv/SeventvEmotes.cpp]: I'd rather avoid making unrelated code changes. I'd want changes in Ffz class reverted.

File: src/providers/seventv/SeventvEmotes.cpp (hunk 7 above)
Comment (zneix) [reply src/providers/seventv/SeventvEmotes.cpp:40]: Fair enough, sorry for my excessive nitpickiness.

File: src/providers/seventv/SeventvEmotes.cpp (hunk 4 above)
Comment (goldbattle) [reply src/providers/seventv/SeventvEmotes.cpp]: See latest.

**Comment by goldbattle** [comment]
Are there any more review comments for me to address? If not, could this be relooked at and merged? Thanks!

**Comment by TroyKomodo** [comment]
@goldbattle Thank you for your effort on this PR, however the v3 of the API is about to be released in the next few weeks and I think we should look to use that API instead of the soon to be deprecated v2 API.

**Comment by TroyKomodo** [comment]
I completely understand if you do not want to take up that task, and i really appreciate your effort on this PR sorry for the delay, we have gone through a few c++ devs now.

**Comment by goldbattle** [comment]
No worries I can take a look once it releases. My goal is to upstream
support.

**Comment by TroyKomodo** [comment]
@goldbattle Also, https://github.com/SevenTV/EventAPI/commit/4c7b52475c5e9f8f8f4f2ee0703374b63a04f22c Added support for websocket on the event api, which means that this could be used in chatterino for auto updating channel emotes. This feature will persist into v3.

**Comment by jammehcow** [comment]
> @goldbattle Also, [SevenTV/EventAPI@4c7b524](https://github.com/SevenTV/EventAPI/commit/4c7b52475c5e9f8f8f4f2ee0703374b63a04f22c) Added support for websocket on the event api, which means that this could be used in chatterino for auto updating channel emotes. This feature will persist into v3.

Only concern I have with this is the lack of event support from other emote providers in chatterino. This would result in inconsistent & confusing behavior between for a new user. I think support for v3 should roughly mirror that of the other providers and we can revisit down the line.

**Comment by AnatoleAM** [comment]
> > @goldbattle Also, [SevenTV/EventAPI@4c7b524](https://github.com/SevenTV/EventAPI/commit/4c7b52475c5e9f8f8f4f2ee0703374b63a04f22c) Added support for websocket on the event api, which means that this could be used in chatterino for auto updating channel emotes. This feature will persist into v3.
> 
> Only concern I have with this is the lack of event support from other emote providers in chatterino. This would result in inconsistent & confusing behavior between for a new user. I think support for v3 should roughly mirror that of the other providers and we can revisit down the line.

Just because other providers have less features shouldn't mean one can't have extra features. Though generally I would hold off on merging this and/or implementing the Event API until v3 is fully released, since it's now essentially in a testing phase before deploy to production
